crates/synthesis/src/linalg/cos_sin_decomp.rs

```diff
@@ -225,21 +172,28 @@ pub fn cos_sin_decomposition(u: DMatrix<Complex64>) -> CosSinDecompReturn {
         .map(|(&ci, &si)| si.atan2(ci))
         .collect();
 
-    (l0, l1, r0, r1, thetas)
+    CosSinDecompReturn {
+        l0,
+        l1,
+        r0,
+        r1,
+        thetas,
+    }
 }
 
-// TODO: Remove this function and all the python interface when QSD is ported to rust
 #[pyfunction]
 pub fn cossin<'py>(py: Python<'py>, u: PyReadonlyArray2<Complex64>) -> PyResult<Bound<'py, PyAny>> {
     let array = u.as_array();
     let shape = array.shape();
-    let mat: DMatrix<Complex64> = DMatrix::from_fn(shape[0], shape[1], |i, j| array[[i, j]]);
-    let res = cos_sin_decomposition(mat);
-    Ok((
-        (res.0.to_pyarray(py), res.1.to_pyarray(py)),
-        res.4.to_pyarray(py),
-        (res.2.to_pyarray(py), res.3.to_pyarray(py)),
-    )
+    let mat: Mat<Complex64> = Mat::from_fn(shape[0], shape[1], |i, j| array[[i, j]]);
+    let res = cos_sin_decomposition(mat.as_ref());
+
+    let arr0 = res.l0.as_ref().into_ndarray().to_pyarray(py);
+    let arr1 = res.l1.as_ref().into_ndarray().to_pyarray(py);
+    let arr2 = res.r0.as_ref().into_ndarray().to_pyarray(py);
+    let arr3 = res.r1.as_ref().into_ndarray().to_pyarray(py);
```

Comment:
It'd really be nice if we had a no-copy path from faer -> numpy here since we have an owned array already and we could just return a numpy array that points to it (using the `IntoPyArray` trait's `.into_pyarray()` method). But we would have to implement the trait ourselves since it's not in rust-numpy.